src/Sulu/Bundle/ContentBundle/Controller/NodeController.php

```diff
@@ -574,48 +583,72 @@ public function filterAction(Request $request)
-        $shadowBaseLanguage = $this->getRequestParameter($request, 'shadowBaseLanguage', null);
+        $type = $request->get('type', 'page');
 
-        $state = $this->getRequestParameter($request, 'state');
-        $type = $request->query->get('type', 'page');
+        try {
+            $document = $this->getDocumentManager()->find(
+                $uuid,
+                $language,
+                ['type' => $type, 'load_ghost_content' => false]
+            );
+        } catch (DocumentNotFoundException $e) {
+            $e = new EntityNotFoundException(PageDocument::class, $uuid, $e);
 
-        if ($state !== null) {
-            $state = intval($state);
+            return $this->handleView($this->view($e->toArray(), 404));
         }
 
+        $webspace = $this->getWebspace($request);
         $data = $request->request->all();
+        $data['structureType'] = $data['template'];
+        $data['shadowLocaleEnabled'] = $this->getRequestParameter($request, 'shadowOn', false);
+        $data['shadowLocale'] = $this->getRequestParameter($request, 'shadowBaseLanguage', null);
+        $data['workflowStage'] = $this->getRequestParameter($request, 'state');
+
+        // disable csrf protection, since we can't produce a token, because the form is cached on the client
+        $options = [
+            'csrf_protection' => false,
+            'webspace_key' => $webspace,
+        ];
 
-        $template = isset($data['template']) ? $data['template'] : null;
+        $form = $this->createForm($type, $document, $options);
+        $form->submit($data, false);
+
+        if (!$form->isValid()) {
+            throw new InvalidFormException($form);
+        }
 
         try {
-            $mapperRequest = ContentMapperRequest::create()
-                ->setType($type)
-                ->setTemplateKey($template)
-                ->setWebspaceKey($webspace)
-                ->setUserId($this->getUser()->getId())
-                ->setState($state)
-                ->setIsShadow($isShadow)
-                ->setShadowBaseLanguage($shadowBaseLanguage)
-                ->setLocale($language)
-                ->setUuid($uuid)
-                ->setData($data);
-            $result = $this->getRepository()->saveNodeRequest($mapperRequest);
-        } catch (MandatoryPropertyException $ex) {
-            return $this->handleView($this->view($ex->getMessage(), 400));
+            $this->getDocumentManager()->persist(
+                $document,
+                $language,
+                [
+                    'user' => $this->getUser()->getId(),
+                    'clear_missing_content' => false,
+                ]
+            );
+            $this->getDocumentManager()->flush();
+        } catch (MandatoryPropertyException $e) {
+            return $this->handleView($this->view($e->getMessage(), 400));
         }
 
-        return $this->handleView(
-            $this->view($result)
+        $this->dispatchContentNodeEvent($document);
```

Comment:
Actually, when thinking about it... There should already be the possibility to do this during the flush, otherwise this only works for actions done via the controller... Probably have to fix that properly now, but only dispatch the old event for now. Later all our subscribers should listen to the real flush event.